wrappers/python/tests/layer/test_gpt2_attention.py

```diff
@@ -0,0 +1,201 @@
+# @version 1.0.0
+
+from dataclasses import dataclass
+
+import numpy as np
+import pytest
+import torch
+from transformers.models.gpt2.modeling_gpt2 import (
+    GPT2Attention as GPT2Attention_torch, GPT2Config as GPT2Config_torch)
+
+import nntile
+from nntile.model.gpt2_config import GPT2ConfigNNTile
+from nntile.tensor import TensorMoments, TensorTraits, Tensor_bool
+from nntile.utils.constructors import to_numpy
+
+# NNTile dtype via corresponding Tensor type
+dtype2nntile = {
+        'fp32': nntile.tensor.Tensor_fp32,
+        'fp32_fast_tf32': nntile.tensor.Tensor_fp32_fast_tf32,
+        'bf16': nntile.tensor.Tensor_bf16,
+}
+
+dtype2tol = {
+        'fp32': {'rtol': 1e-6},
+        'fp32_fast_tf32': {'rtol': 6e-4},
+        'bf16': {'rtol': 1.6e-2},
+}
+
+nocuda = pytest.mark.skipif(not torch.cuda.is_available(), reason='no cuda')
+
+
+@dataclass
+class GPT2AttentionTestParams:
+    n_emb: int
+    n_emb_tile: int
+    n_seq: int
+    n_seq_tile: int
+    n_batch: int
+    n_batch_tile: int
+    n_head: int
+    n_head_tile: int
+    layer_idx: int = 0
+
+
+single_tile = GPT2AttentionTestParams(
+    n_emb=128,
+    n_emb_tile=128,
+    n_seq=64,
+    n_seq_tile=64,
+    n_batch=3,
+    n_batch_tile=3,
+    n_head=4,
+    n_head_tile=4,
+)
+
+multiple_tiles = GPT2AttentionTestParams(
+    n_emb=128,
+    n_emb_tile=32,
+    n_seq=64,
+    n_seq_tile=16,
+    n_batch=4,
+    n_batch_tile=1,
+    n_head=16,
+    n_head_tile=8,
+)
+
+
+def generate_inputs(dtype: str, params: GPT2AttentionTestParams):
+    rng = np.random.default_rng(42)
+    torch_layer_config = GPT2Config_torch(
+        n_embd=params.n_emb,
+        n_head=params.n_head,
+        use_cache=False,
+        resid_pdrop=0.0,
+        embd_pdrop = 0.0,
+        attn_pdrop=0.0,
+        scale_attn_weights = True
+    )
+
+    nntile_config = GPT2ConfigNNTile(
+        vocab_size=torch_layer_config.vocab_size,
+        vocab_embed_dim_tile=params.n_emb,
+        hidden_size=params.n_emb,
+        hidden_size_tile=params.n_emb_tile,
+        intermediate_size=torch_layer_config.n_inner,
+        intermediate_size_tile=torch_layer_config.n_inner,
+        n_head=params.n_head,
+        n_head_tile=params.n_head_tile,
+        dtype=dtype
+    )
+
+    torch_layer = GPT2Attention_torch(
+        torch_layer_config, is_cross_attention=False, layer_idx=params.layer_idx
+    )
+
+    x_shape = [params.n_emb, params.n_seq, params.n_batch]
+    x_basetile = [params.n_emb_tile, params.n_seq_tile, params.n_batch_tile]
+    x_type = dtype2nntile[dtype]
+
+    x_q_traits = TensorTraits(x_shape, x_basetile)
+    x_q_distr = [0] * x_q_traits.grid.nelems
+    x_value = x_type(x_q_traits, x_q_distr, 0)
+    x_grad = x_type(x_q_traits, x_q_distr, 0)
+    X = TensorMoments(x_value, x_grad, grad_required=True)
+
+    x_random = rng.standard_normal(x_shape)
+    x_nntile = np.array(x_random, dtype=np.float32, order="F")
+    x_value.from_array(x_nntile)
+    x_torch = torch.Tensor(x_nntile.T)
```

Comment:
add `.requires_grad_(True)` as it will be needed to check gradient over input in an updated `test_backward`